File: engine/movegen.py
"""Move generation.

``generate_pseudo_moves`` produces every move that follows the movement
rules of the pieces but may leave the mover's own king in check.
``generate_legal_moves`` filters those with make/unmake and appends
한수쉼 (PASS_MOVE) when the side to move is not in check.

Rules encoded here:
  * 차 slides orthogonally; inside a palace it may also slide along the
    marked diagonals (corner - centre - corner).
  * 포 moves orthogonally (and along palace diagonals) by jumping exactly one
    screen piece.  The screen may not be a cannon (either colour) and a
    cannon may never capture a cannon.
  * 마 : one orthogonal step then one diagonal step outward; the orthogonal
    square must be empty.
  * 상 : one orthogonal step then two diagonal steps outward; both
    intermediate squares must be empty.
  * 졸/병 : one step forward or sideways, never backward; forward diagonals
    along palace lines when standing in a palace.
  * 궁/사 : one step along any palace line, confined to the palace.
  * 한수쉼 (pass) is legal whenever the side to move is not in check.
"""
from __future__ import annotations

from .board import (
    CANNON,
    CHARIOT,
    ELEPHANT,
    GUARD,
    HORSE,
    KING,
    NUM_SQUARES,
    PASS_MOVE,
    PAWN,
    Position,
)
from .geometry import (
    ELEPHANT_MOVES,
    HORSE_MOVES,
    PALACE_STEPS,
    PAWN_MOVES,
    RAYS_ALL,
)

_SLIDER = CHARIOT
_JUMPER = CANNON


def generate_pseudo_moves(pos: Position, captures_only: bool = False) -> list[int]:
    board = pos.board
    side = pos.side
    moves: list[int] = []
    push = moves.append
    pawn_tbl = PAWN_MOVES[side]

    for frm in range(NUM_SQUARES):
        p = board[frm]
        if not p or (p >> 3) != side:
            continue
        t = p & 7
        base = frm * NUM_SQUARES

        if t == CHARIOT:
            for ray in RAYS_ALL[frm]:
                for to in ray:
                    q = board[to]
                    if not q:
                        if not captures_only:
                            push(base + to)
                    else:
                        if (q >> 3) != side:
                            push(base + to)
                        break

        elif t == CANNON:
            for ray in RAYS_ALL[frm]:
                screen = False
                for to in ray:
                    q = board[to]
                    if not screen:
                        if q:
                            if (q & 7) == CANNON:
                                break
                            screen = True
                    elif not q:
                        if not captures_only:
                            push(base + to)
                    else:
                        if (q >> 3) != side and (q & 7) != CANNON:
                            push(base + to)
                        break

        elif t == HORSE:
            for block, to in HORSE_MOVES[frm]:
                if board[block]:
                    continue
                q = board[to]
                if not q:
                    if not captures_only:
                        push(base + to)
                elif (q >> 3) != side:
                    push(base + to)

        elif t == ELEPHANT:
            for b1, b2, to in ELEPHANT_MOVES[frm]:
                if board[b1] or board[b2]:
                    continue
                q = board[to]
                if not q:
                    if not captures_only:
                        push(base + to)
                elif (q >> 3) != side:
                    push(base + to)

        elif t == PAWN:
            for to in pawn_tbl[frm]:
                q = board[to]
                if not q:
                    if not captures_only:
                        push(base + to)
                elif (q >> 3) != side:
                    push(base + to)

        else:  # KING or GUARD
            for to in PALACE_STEPS[frm]:
                q = board[to]
                if not q:
                    if not captures_only:
                        push(base + to)
                elif (q >> 3) != side:
                    push(base + to)

    return moves


def is_legal_after_make(pos: Position, mover: int) -> bool:
    """Call right after ``pos.make_move``: was the move legal for ``mover``?"""
    return not pos.attacked_by(pos.king_sq[mover], mover ^ 1)


def _must_resolve_bikjang(pos: Position) -> bool:
    """Forced-bikjang variant: the side to move faces a 빅장 and is not in check."""
    return pos.forced_bikjang and pos.bikjang and not pos.in_check()


def generate_legal_moves(pos: Position, include_pass: bool = True) -> list[int]:
    side = pos.side
    legal: list[int] = []
    must_resolve = _must_resolve_bikjang(pos)
    for m in generate_pseudo_moves(pos):
        pos.make_move(m)
        ok = not pos.attacked_by(pos.king_sq[side], side ^ 1)
        if ok and must_resolve and pos.kings_facing():
            ok = False          # leaving the kings facing is not a legal answer to a 빅장
        pos.unmake_move()
        if ok:
            legal.append(m)
    if include_pass and not pos.in_check():
        legal.append(PASS_MOVE)     # under a forced 빅장 the pass accepts the ending
    return legal


def is_legal_move(pos: Position, m: int) -> bool:
    if m == PASS_MOVE:
        return not pos.in_check()
    return m in generate_pseudo_moves(pos) and m in generate_legal_moves(pos, include_pass=False)


def gives_check(pos: Position, m: int) -> bool:
    pos.make_move(m)
    chk = pos.in_check()
    pos.unmake_move()
    return chk


def perft(pos: Position, depth: int, include_pass: bool = False, stop_at_game_end: bool = False) -> int:
    """Count leaf nodes of the legal move tree (pass moves optional).

    ``stop_at_game_end`` treats positions ended by rule (two consecutive passes,
    an accepted/unresolved 빅장) as leaves, which is how Fairy-Stockfish counts;
    with ``include_pass`` and a forced-bikjang position this reproduces its
    janggi perft numbers exactly.
    """
    if depth == 0:
        return 1
    if stop_at_game_end and (pos.passes >= 2 or (pos.bikjang and pos.prev_bikjang())):
        return 0
    side = pos.side
    total = 0
    pseudo = generate_pseudo_moves(pos)
    must_resolve = _must_resolve_bikjang(pos)
    if depth == 1 and not stop_at_game_end and not must_resolve:
        for m in pseudo:
            pos.make_move(m)
            if not pos.attacked_by(pos.king_sq[side], side ^ 1):
                total += 1
            pos.unmake_move()
        if include_pass and not pos.in_check():
            total += 1
        return total
    for m in pseudo:
        pos.make_move(m)
        if not pos.attacked_by(pos.king_sq[side], side ^ 1) and not (must_resolve and pos.kings_facing()):
            total += perft(pos, depth - 1, include_pass, stop_at_game_end)
        pos.unmake_move()
    if include_pass and not pos.in_check():
        pos.make_move(PASS_MOVE)
        total += perft(pos, depth - 1, include_pass, stop_at_game_end)
        pos.unmake_move()
    return total

File: android/game/src/main/kotlin/com/example/janggiai/game/MoveGen.kt
package com.example.janggiai.game

import com.example.janggiai.game.Board.NUM_SQUARES
import com.example.janggiai.game.Piece.CANNON
import com.example.janggiai.game.Piece.CHARIOT
import com.example.janggiai.game.Piece.ELEPHANT
import com.example.janggiai.game.Piece.HORSE
import com.example.janggiai.game.Piece.PAWN

/**
 * Move generation (port of `engine/movegen.py`).
 *
 *  * 차 slides orthogonally, plus palace diagonals.
 *  * 포 jumps exactly one screen (never a cannon) and never captures a cannon.
 *  * 마 one orthogonal + one diagonal step (first square must be empty); 상 one orthogonal + two
 *    diagonal steps (both intermediate squares empty).
 *  * 졸/병 forward or sideways, palace diagonals forward inside a palace.
 *  * 궁/사 one step along palace lines.
 *  * 한수쉼 ([Move.PASS]) is legal whenever the side to move is not in check.
 */
object MoveGen {
    fun pseudoMoves(pos: Position, capturesOnly: Boolean = false): IntArrayList {
        val board = pos.board
        val side = pos.side
        val moves = IntArrayList(64)
        val pawnTbl = Geometry.PAWN_MOVES[side]
        for (from in 0 until NUM_SQUARES) {
            val p = board[from]
            if (p == 0 || Piece.colorOf(p) != side) continue
            val base = from * NUM_SQUARES
            when (Piece.typeOf(p)) {
                CHARIOT -> for (ray in Geometry.RAYS_ALL[from]) {
                    for (to in ray) {
                        val q = board[to]
                        if (q == 0) { if (!capturesOnly) moves.add(base + to) }
                        else { if (Piece.colorOf(q) != side) moves.add(base + to); break }
                    }
                }
                CANNON -> for (ray in Geometry.RAYS_ALL[from]) {
                    var screen = false
                    for (to in ray) {
                        val q = board[to]
                        if (!screen) {
                            if (q != 0) { if (Piece.typeOf(q) == CANNON) break; screen = true }
                        } else if (q == 0) {
                            if (!capturesOnly) moves.add(base + to)
                        } else {
                            if (Piece.colorOf(q) != side && Piece.typeOf(q) != CANNON) moves.add(base + to)
                            break
                        }
                    }
                }
                HORSE -> for (hm in Geometry.HORSE_MOVES[from]) {
                    if (board[hm[0]] != 0) continue
                    val q = board[hm[1]]
                    if (q == 0) { if (!capturesOnly) moves.add(base + hm[1]) }
                    else if (Piece.colorOf(q) != side) moves.add(base + hm[1])
                }
                ELEPHANT -> for (em in Geometry.ELEPHANT_MOVES[from]) {
                    if (board[em[0]] != 0 || board[em[1]] != 0) continue
                    val q = board[em[2]]
                    if (q == 0) { if (!capturesOnly) moves.add(base + em[2]) }
                    else if (Piece.colorOf(q) != side) moves.add(base + em[2])
                }
                PAWN -> for (to in pawnTbl[from]) {
                    val q = board[to]
                    if (q == 0) { if (!capturesOnly) moves.add(base + to) }
                    else if (Piece.colorOf(q) != side) moves.add(base + to)
                }
                else -> for (to in Geometry.PALACE_STEPS[from]) {  // KING / GUARD
                    val q = board[to]
                    if (q == 0) { if (!capturesOnly) moves.add(base + to) }
                    else if (Piece.colorOf(q) != side) moves.add(base + to)
                }
            }
        }
        return moves
    }

    /** Forced-bikjang variant: the side to move faces a 빅장 and is not in check. */
    fun mustResolveBikjang(pos: Position): Boolean = pos.forcedBikjang && pos.bikjang && !pos.inCheck()

    fun legalMoves(pos: Position, includePass: Boolean = true): List<Int> {
        val side = pos.side
        val legal = ArrayList<Int>(48)
        val mustResolve = mustResolveBikjang(pos)
        val pseudo = pseudoMoves(pos)
        for (i in 0 until pseudo.size) {
            val m = pseudo[i]
            pos.makeMove(m)
            var ok = !pos.attackedBy(pos.kingSq[side], side xor 1)
            if (ok && mustResolve && pos.kingsFacing()) ok = false   // leaving the kings facing is not an answer to a 빅장
            pos.unmakeMove()
            if (ok) legal += m
        }
        if (includePass && !pos.inCheck()) legal += Move.PASS   // under a forced 빅장 the pass accepts the ending
        return legal
    }

    fun isLegal(pos: Position, m: Int): Boolean =
        if (m == Move.PASS) !pos.inCheck() else legalMoves(pos, includePass = false).contains(m)

    fun givesCheck(pos: Position, m: Int): Boolean {
        pos.makeMove(m)
        val chk = pos.inCheck()
        pos.unmakeMove()
        return chk
    }

    /**
     * Leaf count of the legal move tree.  [stopAtGameEnd] treats rule-ended positions (two passes,
     * unresolved/accepted 빅장) as leaves, which is how Fairy-Stockfish counts; with [includePass]
     * on a forced-bikjang position this reproduces its `go perft` numbers exactly.
     */
    fun perft(pos: Position, depth: Int, includePass: Boolean = false, stopAtGameEnd: Boolean = false): Long {
        if (depth == 0) return 1
        if (stopAtGameEnd && (pos.passes >= 2 || (pos.bikjang && pos.prevBikjang()))) return 0
        val side = pos.side
        var total = 0L
        val pseudo = pseudoMoves(pos)
        val mustResolve = mustResolveBikjang(pos)
        if (depth == 1 && !stopAtGameEnd && !mustResolve) {
            for (i in 0 until pseudo.size) {
                pos.makeMove(pseudo[i])
                if (!pos.attackedBy(pos.kingSq[side], side xor 1)) total++
                pos.unmakeMove()
            }
            if (includePass && !pos.inCheck()) total++
            return total
        }
        for (i in 0 until pseudo.size) {
            pos.makeMove(pseudo[i])
            if (!pos.attackedBy(pos.kingSq[side], side xor 1) && !(mustResolve && pos.kingsFacing()))
                total += perft(pos, depth - 1, includePass, stopAtGameEnd)
            pos.unmakeMove()
        }
        if (includePass && !pos.inCheck()) {
            pos.makeMove(Move.PASS)
            total += perft(pos, depth - 1, includePass, stopAtGameEnd)
            pos.unmakeMove()
        }
        return total
    }
}

/** Minimal growable int list (avoids boxing in the hot move-generation path). */
class IntArrayList(capacity: Int = 16) {
    private var data = IntArray(capacity)
    var size = 0
        private set

    fun add(v: Int) {
        if (size == data.size) data = data.copyOf(data.size * 2)
        data[size++] = v
    }

    operator fun get(i: Int): Int = data[i]
    fun toList(): List<Int> = List(size) { data[it] }
}
